src/3d/symbols/qgslinevertexdata_p.cpp

```diff
@@ -0,0 +1,152 @@
+}
+
+QByteArray QgsLineVertexData::createIndexBuffer()
+{
+  QByteArray indexBufferData;
+  indexBufferData.resize( indexes.size() * sizeof( int ) );
+  unsigned int *rawIndexArray = reinterpret_cast<unsigned int *>( indexBufferData.data() );
+  int idx = 0;
+  for ( unsigned int indexVal : qgis::as_const( indexes ) )
+  {
+    rawIndexArray[idx++] = indexVal;
+  }
+  return indexBufferData;
+}
+
+Qt3DRender::QGeometry *QgsLineVertexData::createGeometry( Qt3DCore::QNode *parent )
+{
+  Qt3DRender::QBuffer *vertexBuffer = new Qt3DRender::QBuffer( Qt3DRender::QBuffer::VertexBuffer, parent );
+  vertexBuffer->setData( createVertexBuffer() );
+
+  Qt3DRender::QBuffer *indexBuffer = new Qt3DRender::QBuffer( Qt3DRender::QBuffer::IndexBuffer, parent );
+  indexBuffer->setData( createIndexBuffer() );
+
+  QgsDebugMsg( QString( "vertex buffer %1 MB  index buffer %2 MB " ).arg( vertexBuffer->data().count() / 1024. / 1024. ).arg( indexBuffer->data().count() / 1024. / 1024. ) );
+
+  Qt3DRender::QAttribute *positionAttribute = new Qt3DRender::QAttribute( parent );
+  positionAttribute->setAttributeType( Qt3DRender::QAttribute::VertexAttribute );
+  positionAttribute->setBuffer( vertexBuffer );
+  positionAttribute->setVertexBaseType( Qt3DRender::QAttribute::Float );
+  positionAttribute->setVertexSize( 3 );
+  positionAttribute->setName( Qt3DRender::QAttribute::defaultPositionAttributeName() );
+
+  Qt3DRender::QAttribute *indexAttribute = new Qt3DRender::QAttribute( parent );
+  indexAttribute->setAttributeType( Qt3DRender::QAttribute::IndexAttribute );
+  indexAttribute->setBuffer( indexBuffer );
+  indexAttribute->setVertexBaseType( Qt3DRender::QAttribute::UnsignedInt );
+
+  Qt3DRender::QGeometry *geom = new Qt3DRender::QGeometry;
+  geom->addAttribute( positionAttribute );
+  geom->addAttribute( indexAttribute );
+  return geom;
+}
+
+void QgsLineVertexData::addLineString( const QgsLineString &lineString, float extraHeightOffset )
+{
+  if ( withAdjacency )
+    indexes << vertices.count();  // add the following vertex (for adjacency)
+
+  QgsPoint centroid;
+  if ( altBinding == Qgs3DTypes::AltBindCentroid )
+    centroid = lineString.centroid();
+
+  for ( int i = 0; i < lineString.vertexCount(); ++i )
+  {
+    QgsPoint p = lineString.pointN( i );
+    float z = Qgs3DUtils::clampAltitude( p, altClamping, altBinding, baseHeight + extraHeightOffset, centroid, *mapSettings );
+
+    vertices << QVector3D( p.x() - mapSettings->origin().x(), z, -( p.y() - mapSettings->origin().y() ) );
+    indexes << vertices.count() - 1;
+  }
+
+  if ( withAdjacency )
+    indexes << vertices.count() - 1;  // add the last vertex (for adjacency)
+
+  indexes << 0;  // add primitive restart
+}
+
+
+void QgsLineVertexData::addVerticalLines( const QgsLineString &lineString, float verticalLength )
+{
+  QgsPoint centroid;
+  if ( altBinding == Qgs3DTypes::AltBindCentroid )
+    centroid = lineString.centroid();
+
+  for ( int i = 0; i < lineString.vertexCount(); ++i )
```

Comment:
Didn't you write a vertex iterator?

Reply:
Yes, vertex iterator is good when using arbitrary geometries. But when dealing with a particular type of geometry, why not use something with less overhead :-)

Reply:
I was assuming that iterators have less overhead than repeatedly calling vertexCount() and looking up the item at a given index, but I admit I didn't do any check for ground truth :)